VideoApp/VideoApp/Video/VideoAppCameraSource.swift

```diff
@@ -0,0 +1,168 @@
+        let frameRate: UInt
+        
+        switch appSettingsStore.videoCodec {
+        case .h264, .vp8:
+            // 640 x 480 squarish crop (1.13:1)
+            targetSize = CMVideoDimensions(width: 544, height: 480)
+            
+            cropRatio = CGFloat(targetSize.width) / CGFloat(targetSize.height)
+            frameRate = 20
+        case .vp8Simulcast:
+            // 1024 x 768 squarish crop (1.25:1) on most iPhones. 1280 x 720 squarish crop (1.25:1) on the iPhone X
+            // and models that don't have 1024 x 768.
+            targetSize = CMVideoDimensions(width: 900, height: 720)
+            
+            cropRatio = CGFloat(targetSize.width) / CGFloat(targetSize.height)
+            frameRate = 24 // With simulcast enabled there are 3 temporal layers, allowing a frame rate of f/4
+        }
+        
+        let captureDevice = CameraSource.captureDevice(position: .front)!
+        let preferredFormat = selectVideoFormatBySize(captureDevice: captureDevice, targetSize: targetSize)
+        preferredFormat.frameRate = min(preferredFormat.frameRate, frameRate)
+        
+        let cropDimensions: CMVideoDimensions
+        
+        if preferredFormat.dimensions.width > preferredFormat.dimensions.height {
+            cropDimensions = CMVideoDimensions(
+                width: Int32(CGFloat(preferredFormat.dimensions.height) * cropRatio),
+                height: preferredFormat.dimensions.height
+            )
+        } else {
+            cropDimensions = CMVideoDimensions(
+                width: preferredFormat.dimensions.width,
+                height: Int32(CGFloat(preferredFormat.dimensions.width) * cropRatio)
+            )
+        }
+        
+        let outputFormat = VideoFormat()
+        outputFormat.dimensions = cropDimensions
+        outputFormat.pixelFormat = preferredFormat.pixelFormat
+        outputFormat.frameRate = 0
+        
+        cameraSource.requestOutputFormat(outputFormat)
+        
+        cameraSource.startCapture(device: captureDevice, format: preferredFormat) { [weak self] _, _, error in
+            guard error == nil else { return }
+            
+            self?.localMediaController.videoCaptureStarted()
+        }
+    }
+        
+    private func selectVideoFormatBySize(captureDevice: AVCaptureDevice, targetSize: CMVideoDimensions) -> VideoFormat {
+        let supportedFormats = Array(CameraSource.supportedFormats(captureDevice: captureDevice)) as! [VideoFormat]
+        
+        // Cropping might be used if there is not an exact match
+        for format in supportedFormats {
+            guard
+                format.pixelFormat == .formatYUV420BiPlanarFullRange &&
+                    format.dimensions.width >= targetSize.width &&
+                    format.dimensions.height >= targetSize.height
+                else {
+                    continue
+            }
+            
+            return format
+        }
+        
+        fatalError()
```

Comment:
Is a `fatalError()` necessary here? In the Obj-C code it would return `nil` if no format was found... I'm not sure if that is an actual possible condition or not...

Reply:
The compiler enforces all paths in the function return a value and this is a way around that. I could make the return type optional and return nil but then we just have some other not so great error handling down the line dealing with the optional. So not very good options until significant refactor.

However when I was looking at objc code before how I read it was the function will crash if it doesn't find a result. And so this was equivalent. But I realize now how pointers are handled it probably would have just fizzled out later down the line. Man kind of messes with me going between objc and Swift pointer logic.